### PR TITLE
dev-ruby/http: add 5.1.0

### PR DESCRIPTION
**Pull request by matoro**
Giant push to the Ruby tree to bring up dev-ruby/http to the 5.x branch.
http-5.x changes the underlying C parser from http-parser to llhttp-ffi,
which pulls in most of the socketry async ecosystem for its tests.  All
of these packages have been tested on sparc, using all compatible Ruby
versions from ruby26 to ruby31.

KNOWN ISSUES:

* dev-ruby/async-http tests can hang on ruby27 only
* dev-ruby/dalli tests are RESTRICTed because they spin up a local
  memcached instance, which is broken on sparc (see bug 636552)
* In order to get the benefits of respect-cflags.patch for packages that
  use ffi-compiler (such as dev-ruby/llhttp-ffi), you must rebuild
  dev-ruby/ffi-compiler
* dev-ruby/ffi has two failing tests, but passes everything else.  This
  does not seem to affect functionality.  Documented in bug 847286 and
  upstream https://github.com/ffi/ffi/pull/957
* dev-ruby/simplecov tests fail for ruby3x, already filed by someone
  else as bug 835620.  This may be fixed in newer upstream versions

### CODE REVIEW

**Comment by thesamesam** [comment]
This seems to touch multiple packages, so one commit per package :)

**Comment by matoro** [comment]
Rebased.  `pkgcheck` does not seem to understand the syntax that `ruby_add_xdepend` takes for pinned versions.

**Comment by thesamesam** [comment]
> Rebased. `pkgcheck` does not seem to understand the syntax that `ruby_add_xdepend` takes for pinned versions.

No, I think maybe misunderstanding what * does?

e.g. there's no dev-ruby/console-1.10*, we have dev-ruby/console-1.15 in this PR.

**Comment by matoro** [comment]
> Rebased. pkgcheck does not seem to understand the syntax that ruby_add_xdepend takes for pinned versions.
> 
> No, I think maybe misunderstanding what * does?
> 
> e.g. there's no dev-ruby/console-1.10*, we have dev-ruby/console-1.15 in this PR.

Okay, CI fixed.  I added subslots to everything in order to be able to specify major or minor versions.  Removed some unnecessary older versions as I misunderstood the gemspec tilde operator at first.

**Comment by graaff** [comment]
Reviewing this is going to be a lot of work. I'm not sure when I will be able to do this, so it may take some time.

**Review by graaff — COMMENTED** [review]
Please don't mix different changes into a single commit. Either keyword for ~sparc, or move to EAPI 8. If you move to EAPI 8 then that also requires a new revision to ensure any file changes are properly accounted for and we can handle any bugs properly.

I have keyworded dev-ruby/ast for ~sparc in the gentoo tree, so this commit can be removed.

File: dev-ruby/certificate_authority/certificate_authority-1.0.0.ebuild [hunk 1]
```diff
@@ -1,8 +1,8 @@
 # Copyright 1999-2021 Gentoo Authors
```
Comment (graaff) [inline dev-ruby/certificate_authority/certificate_authority-1.0.0.ebuild:1]: These kinds of changes (EAPI, adding a patch) require a new revision of the ebuild.

**Comment by graaff** [comment]
This pull request is all over the place and that makes it very hard to do something with it. Please create focused pull requests that tackle one thing: add ~sparc keyword, or add ~ruby31. Note that I'm not very enthusiastic at this point about backporting ruby31 changes. Instead try to convince upstream to release a new version.

Leave the EAPI alone: it introduces a lot of extra clutter in all commits and requires new revisions which makes reviewing harder. There really is no pressing need to change this at all.

**Comment by matoro** [comment]
> This pull request is all over the place and that makes it very hard to do something with it. Please create focused pull requests that tackle one thing: add ~sparc keyword, or add ~ruby31. Note that I'm not very enthusiastic at this point about backporting ruby31 changes. Instead try to convince upstream to release a new version.
> 
> Leave the EAPI alone: it introduces a lot of extra clutter in all commits and requires new revisions which makes reviewing harder. There really is no pressing need to change this at all.

I know there's a lot going on here and that it's going to be somewhat difficult whichever way it's approached.  But realistically, will breaking up each of these into individual PRs actually be looked at in a timely manner?  Splitting out ruby31 from keywording will increase that count even further, probably 70+ PRs.  All of this is mandatory to satisfy pkgcheck if the new version of dev-ruby/http supports ruby31, which it does and should.

File: dev-ruby/http/http-5.0.4.ebuild [hunk 2]
```diff
@@ -0,0 +1,52 @@
+# Copyright 1999-2021 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+USE_RUBY="ruby27 ruby30 ruby31"
+
+RUBY_FAKEGEM_RECIPE_TEST="rspec3"
+RUBY_FAKEGEM_EXTRADOC="CHANGES.md README.md"
+RUBY_FAKEGEM_GEMSPEC="${PN}.gemspec"
+# disabled for now.  pulls in another large dep tree for ruby31
```
Comment (graaff) [inline dev-ruby/http/http-5.0.4.ebuild:10]: yard now have a ruby31 target. We also should not comment out things like this but wait for these packages to be ready for ruby31 before adding it here.

File: dev-ruby/http/http-5.0.4.ebuild [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+# RUBY_FAKEGEM_RECIPE_DOC="yard"
+inherit ruby-fakegem
+
+DESCRIPTION="An easy-to-use client library for making requests from Ruby"
+HOMEPAGE="https://github.com/tarcieri/http"
+
+LICENSE="MIT"
+SLOT="5"
```
Comment (graaff) [inline dev-ruby/http/http-5.0.4.ebuild:18]: It's better to make slots dynamic: e.g. `SLOT="$(ver_cut 1)"`

File: dev-ruby/http/http-5.0.4.ebuild [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+ruby_add_rdepend ">=dev-ruby/addressable-2.8:2
+	>=dev-ruby/http-cookie-1.0:1
+	>=dev-ruby/http-form_data-2.2:2
+	>=dev-ruby/llhttp-ffi-0.4.0:0/0.4"
+
+ruby_add_bdepend "
+	test? (
+		>=dev-ruby/certificate_authority-1.0:1
+		dev-ruby/rspec-its
+		dev-ruby/simplecov
```
Comment (graaff) [inline dev-ruby/http/http-5.0.4.ebuild:31]: In general we try to avoid developer dependencies such as coverage to keep the dependency trees more manageable. We are only interested in the tests. So the preference here is to patch out things like coverage.

**Comment by matoro** [comment]
Alright, I think we're good now.  I ended up keeping the ruby31 work since I really didn't want to throw it away, and the majority of the problems where coming from the USE=doc revdep tree.  Upstream just released 5.1.0 so I bumped to that and included a minor fix I've already submitted upstream.  Here's the breakdown:

STUFF WITH PATCHES (REVBUMPED ALL OF THESE):
* dev-ruby/certificate_authority
* dev-ruby/ffi
* dev-ruby/ffi-compiler
* dev-ruby/sinatra

KEYWORDED:
* dev-ruby/cliver
* dev-ruby/msgpack
* dev-ruby/parser
* dev-ruby/timers

ENABLED ruby31:
* dev-ruby/certificate_authority
* dev-ruby/parser
* dev-ruby/rack-test
* dev-ruby/sinatra

Everything else is brand new.  I hope that this is manageable enough to review as a group?  @graaff

**Comment by graaff** [comment]
I've cherry-picked the keywording commits and these have now been added to the gentoo tree. I expect to review the changes to existing packages next.

File: dev-ruby/async-io/async-io-1.33.0.ebuild [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+SLOT="$(ver_cut 1)/$(ver_cut 1-2)"
+KEYWORDS="~amd64 ~sparc"
+IUSE=""
+
+ruby_add_rdepend "dev-ruby/async"
+
+ruby_add_bdepend "test? (
+	>=dev-ruby/async-container-0.15:0
+	>=dev-ruby/async-rspec-1.10:1
+	dev-ruby/covered
```
Comment (graaff) [inline dev-ruby/async-io/async-io-1.33.0.ebuild:26]: We try to avoid adding developer-only dependencies to packages because they bring a net negative value (there is no value for Gentoo, unlike running tests, but we still need to maintain this and this may bring blockers or additional dependency issues in the future). Please disable or patch out the use of covered and avoid the dependency.

Comment (matoro) [reply dev-ruby/async-io/async-io-1.33.0.ebuild]: Done.  This actually allows me to drop the following packages:

* dev-ruby/async-http (this actually gets rid of the one test hang)
* dev-ruby/async-pool
* dev-ruby/async-rest
* dev-ruby/protocol-hpack
* dev-ruby/protocol-http1
* dev-ruby/protocol-http2
* dev-ruby/protocol-http
* dev-ruby/thread-local
* dev-ruby/traces
* dev-ruby/trenni

File: dev-ruby/certificate_authority/certificate_authority-1.0.0-r1.ebuild [hunk 6]
```diff
@@ -0,0 +1,38 @@
+# Copyright 1999-2021 Gentoo Authors
```
Comment (graaff) [inline dev-ruby/certificate_authority/certificate_authority-1.0.0-r1.ebuild:1]: There has been an upstream release with this commit and I've added that, so this commit is no longer needed.

File: dev-ruby/async/async-1.30.1.ebuild [hunk 7]
```diff
@@ -0,0 +1,40 @@
+# Copyright 2022 Gentoo Authors
```
Comment (graaff) [inline dev-ruby/async/async-1.30.1.ebuild:1]: Why did you add this old slot? It doesn't look like it is needed for any of the packages in this PR, so it will be obsolete from the start. But perhaps I'm overlooking something?

File: dev-ruby/ffi-compiler/files/respect-cflags.patch [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+--- a/lib/ffi-compiler/compile_task.rb
++++ b/lib/ffi-compiler/compile_task.rb
+@@ -26,9 +26,9 @@ module FFI
+         @libraries = []
+         @headers = []
+         @functions = []
+-        @cflags = DEFAULT_CFLAGS.dup
+-        @cxxflags = DEFAULT_CFLAGS.dup
+-        @ldflags = DEFAULT_LDFLAGS.dup
++        @cflags = ENV['CFLAGS'].split || DEFAULT_CFLAGS.dup
```
Comment (graaff) [inline dev-ruby/ffi-compiler/files/respect-cflags.patch:12]: I would like to avoid these kinds of patches to keep closer to upstream and to avoid surprises when people approach this code from the Ruby eco-system side. It's probably best to coordinate these things with upstream instead and see if they are open for this kind of change.

Note that this code is also broken: it will break when no such environment variable is provided.

Comment (matoro) [reply dev-ruby/ffi-compiler/files/respect-cflags.patch:12]: Upstream is dead for ffi-compiler, otherwise I would have done that.  Sorry, I don't know enough Ruby to get this correct then - how should this be in order to work as intended?

Comment (matoro) [reply dev-ruby/ffi-compiler/files/respect-cflags.patch]: Fixed the patch.  My apologies on this.

Comment (graaff) [reply dev-ruby/ffi-compiler/files/respect-cflags.patch]: The most recent change upstream is 4 months ago, so I'm not sure this should be considered dead.

File: dev-ruby/fiber-local/fiber-local-1.0.0.ebuild [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+RUBY_FAKEGEM_EXTRADOC="README.md"
+RUBY_FAKEGEM_GEMSPEC="${PN}.gemspec"
+inherit ruby-fakegem
+
+DESCRIPTION="Provides a class-level mixin to make fiber local state easy"
+HOMEPAGE="https://github.com/socketry/fiber-local"
+SRC_URI="https://github.com/socketry/fiber-local/archive/refs/tags/v${PV}.tar.gz -> ${P}.tar.gz"
+
+LICENSE="MIT"
+SLOT="$(ver_cut 1)/$(ver_cut 1-2)"
```
[the remark this answers is not in the document]
Comment (graaff) [reply dev-ruby/fiber-local/fiber-local-1.0.0.ebuild:17]: I'm not sure what you mean here? Can you explain the reasoning in more detail? What you you mean with `^fiber-local-1.0.0`? I'm not sure what the carrot does in this example because it is not ruby or gentoo syntax for dependencies. I'm also not sure what you would expect to happen with the subslot.

Note that they are not harmless. They bring another thing to take into account and they clutter up overviews of packages so they make maintenance more difficult. I want to understand the benefits before accepting those costs.

[the remark this answers is not in the document]
Comment (matoro) [reply dev-ruby/fiber-local/fiber-local-1.0.0.ebuild:17]: Bleh sorry was working on other languages, I meant `gem 'fiber-local', '~> 1.0.0`.  I did this specifically based on our IRC conversation:

```
2022-06-19 01:56:38     @graaff matoro: I see now that you also change slots. That can't be done without a slotmove, not sure if that is worth the effort.
2022-06-19 01:57:36     matoro  don't have a choice, new packages specify pinned versions
2022-06-19 01:58:37     @graaff I understand that for e.g. a dependency on ~> foo-3.0 it is more communicative to write foo:3 than foo:0 (if the slot was never set) but the latter works fine.
2022-06-19 02:00:53     matoro  for the slots, what to do if we have a foo-3.0 and foo-4.0 in tree and the package needs ~> foo-3.0
2022-06-19 02:01:04     matoro  and both foo-3.0 and foo-4.0 have SLOT=0
2022-06-19 02:01:35     @graaff matoro: in that case we can either do a slotmove or use something like =foo-3*
2022-06-19 02:02:43     @graaff But I don't think there are many packages left where this happens. In the past I tried to be "smart" and avoid slots when the changelog seemed to suggest there were no breaking changes but in the end that appoach just caused more problems.
2022-06-19 02:02:56     matoro  ok one more, this one was the reason I started the SLOT stuff originally - what about if the package specifies ~> foo-3.6 and wee have 3.11 in tree
2022-06-19 02:03:12     matoro  so what that needs is >=foo-3.6:3
2022-06-19 02:03:29     matoro  can't just do >=foo-3.6 because that will match foo-4.x
2022-06-19 02:04:31     @graaff =foo-3* >=foo-3.6
2022-06-19 02:04:51     @graaff But for packages where this actually happens we should probably consider fixing the slot.
```

So because of this conversation, I made sure that all the new ebuilds I added have a SLOT, so that they won't require a slotmove in the future were something to specify a restrictive dependency.

[the remark this answers is not in the document]
Comment (graaff) [reply dev-ruby/fiber-local/fiber-local-1.0.0.ebuild:17]: My question is about the subslot, not the slot itself.

[the remark this answers is not in the document]
Comment (graaff) [reply dev-ruby/fiber-local/fiber-local-1.0.0.ebuild:17]: We don't use subslots anywhere in dev-ruby. Can you explain how you think this helps with specific situations that may arise with dev-ruby packages.

The http gem depends on  llhttp-ffi ~> 0.4.0, so the correct way to deal with that is to use the "0.4" slot here and refer to that in the http ebuild. I don't see how a subslot is going to help here.

[the remark this answers is not in the document]
Comment (matoro) [reply dev-ruby/fiber-local/fiber-local-1.0.0.ebuild:17]: There was an instance in the original PR (I don't recall which, looks like it's gone now) where there were two reverse dependencies with different strictness levels.  One depended on `~> 1.0.0` and the other on `~> 1.0`.

[the remark this answers is not in the document]
Comment (graaff) [reply dev-ruby/fiber-local/fiber-local-1.0.0.ebuild:17]: Yes, that happens sometimes because semantic versioning is not enforced in rubygems. In this case there are two options: a) (preferred) use the "1.0" slot for both packages and if newer minor versions are released (e.g. 1.1.0) make the package with ~> 1.0 depend on both versions with "|| ( )", or b) tweak the gemspec of the package depending on ~> 1.0.0 to be more lenient. 

It is not clear to me how this relates to the subslot or what purpose a subslot would have.

[the remark this answers is not in the document]
Comment (matoro) [reply dev-ruby/fiber-local/fiber-local-1.0.0.ebuild:17]: With a subslot aren't we able to make the first package depend on `dev-ruby/example:1/1.0` and the second package depend on `dev-ruby/example:1`?  This seems much cleaner to me than having to change reverse dependencies on a new release or change what the original author specified.

[the remark this answers is not in the document]
Comment (graaff) [reply dev-ruby/fiber-local/fiber-local-1.0.0.ebuild:17]: That approach has the same drawback as using a dependency like "=dev-ruby/example-1.0*". It is hard to track this when working on the package itself, and consequently easy to break the dependency tree when cleaning up old versions in slots. I'd prefer to have that explicit by using slots only.

[the remark this answers is not in the document]
Comment (matoro) [reply dev-ruby/fiber-local/fiber-local-1.0.0.ebuild:17]: Isn't that what pkgcheck is for?  To yell about dropping versions that are still required by something?

So which is preferred:
* Set `SLOT=1`, package A depends on `dev-ruby/example:1`, package B depends on `=dev-ruby/example-1.0*`
* or, set `SLOT=1.0`, package A depends on `=dev-ruby/example-1*`, package B depends on `dev-ruby/example:1.0`

[the remark this answers is not in the document]
Comment (matoro) [reply dev-ruby/fiber-local/fiber-local-1.0.0.ebuild:17]: @graaff ping on this

**Comment by matoro** [comment]
Hey @graaff , another ping here.

Is the subslots implementation a blocker for you?  I understand that currently Ruby packages are using wildcards in dependency lists, but I've outlined why it's reasonable to use subslots instead, or in addition.  `pkgcheck` should eliminate any consistency issues.

My ffi-compiler PR has been sitting untouched for 30 days now.  Is that sufficient to consider upstream dead and unlikely to commit it, much less create a new release?

Hoping to get some movement on this.  Thanks!

File: dev-ruby/fiber-local/fiber-local-1.0.0.ebuild (hunk 9 above)
[the remark this answers is not in the document]
Comment (thesamesam) [reply dev-ruby/fiber-local/fiber-local-1.0.0.ebuild:17]: I see examples of both in tree. Which do you want matoro to pursue, @graaff?

**Comment by thesamesam** [comment]
I think matoro's been pretty patient here. @graaff could you explain the remaining blockers so I can work on them with matoro? I don't want this to bitrot and it's at risk of that. It looks like the only thing needed is clarification on which slot style you prefer: https://github.com/gentoo/gentoo/pull/25906#discussion_r950985811.

**Comment by matoro** [comment]
Hi @graaff , just one more ping on here.  Could you please take a look at my questions on the two discussions above?

**Comment by thesamesam** [comment]
Ping.

**Comment by matoro** [comment]
Hey @graaff @thesamesam , this is starting to rot.  I've gone over and bumped where possible.  Added the following bumps:
* dev-ruby/sus: bumped from 0.6.2 -> 0.13.0
* dev-ruby/rack-test: tests regressed on sparc at some point.  Filed a ticket and the maintainer was instant in addressing it, see https://github.com/rack/rack-test/issues/323.  This is backported now.
* dev-ruby/sinatra: upstream delayed https://github.com/sinatra/sinatra/pull/1519 to 3.0.0, so the new 2.2.1 version in tree still needs it for ruby31.
* dev-ruby/rspec-files: bumped from 1.1.1 -> 1.1.2
* dev-ruby/async: bumped from 1.30.1 -> 1.30.3 and from 2.0.2 -> 2.1.0
* dev-ruby/async-io: bumped from 1.33.0 -> 1.34.0

Despite being removed from tree I've left ruby26 in since pkgcheck doesn't complain like it does with obsolete python and other packages still in tree also seem to have obsolete USE_RUBY.  I'll leave these to a treewide cleanup later.

Can this please be looked at again?

**Comment by mattst88** [comment]
@graaff: You haven't responded since July and matoro and Sam have pinged you *seven* times since then. With this in mind, I'm going to merge this next week unless you resume reviewing.
